politeiad/backendv2/tstorebe/plugins/plugins.go

```diff
@@ -0,0 +1,221 @@
+// Copyright (c) 2020-2021 The Decred developers
+// Use of this source code is governed by an ISC
+// license that can be found in the LICENSE file.
+
+package plugins
+
+import (
+	"errors"
+
+	backend "github.com/decred/politeia/politeiad/backendv2"
+	"github.com/decred/politeia/politeiad/backendv2/tstorebe/store"
+)
+
+var (
+	// ErrDuplicateBlob is returned when a plugin attempts to save a
+	// blob whose contents are an exact duplicate of an existing blob.
+	ErrDuplicateBlob = errors.New("duplicate blob")
+)
+
+// HookT represents a plugin hook.
+type HookT int
+
+const (
+	// HookTypeInvalid is an invalid plugin hook.
+	HookTypeInvalid HookT = 0
+
+	// HootTypeNewRecordPre is called before a new record is saved to
+	// disk.
+	HookTypeNewRecordPre HookT = 1
+
+	// HootTypeNewRecordPost is called after a new record is saved to
+	// disk.
+	HookTypeNewRecordPost HookT = 2
+
+	// HookTypeEditRecordPre is called before a record update is saved
+	// to disk.
+	HookTypeEditRecordPre HookT = 3
+
+	// HookTypeEditRecordPost is called after a record update is saved
+	// to disk.
+	HookTypeEditRecordPost HookT = 4
+
+	// HookTypeEditMetadataPre is called before a metadata update is
+	// saved to disk.
+	HookTypeEditMetadataPre HookT = 5
+
+	// HookTypeEditMetadataPost is called after a metadata update is
+	// saved to disk.
+	HookTypeEditMetadataPost HookT = 6
+
+	// HookTypeSetRecordStatusPre is called before a record status
+	// change is saved to disk.
+	HookTypeSetRecordStatusPre HookT = 7
+
+	// HookTypeSetRecordStatusPost is called after a record status
+	// change is saved to disk.
+	HookTypeSetRecordStatusPost HookT = 8
+
+	// HookTypePluginPre is called before a plugin command is executed.
+	HookTypePluginPre HookT = 9
+
+	// HookTypePluginPost is called after a plugin command is executed.
+	HookTypePluginPost HookT = 10
```

Comment:
We should have a last so that we can verify that the map has at least the same count as the conts.